test/e2e/podtimingcontroller/podtimingcontroller.go

```diff
@@ -0,0 +1,90 @@
+//go:build e2e
+// +build e2e
+
+package podtimingcontroller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/prometheus/client_golang/prometheus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+)
+
+func SetupWithManager(mgr ctrl.Manager, log logr.Logger, artifactDir string) error {
+	r := &podTimingReconciler{
+		client: mgr.GetClient(),
+		podEnterPhaseDurationSeconds: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "e2e_pod_enter_phase_seconds",
+		}, []string{"phase", "namespace", "name", "uid"}),
+		alreadyRecorded: sets.String{},
+	}
+	if err := prometheus.Register(r.podEnterPhaseDurationSeconds); err != nil {
+		return fmt.Errorf("failed to register e2e_pod_enter_phase_seconds metric: %w", err)
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Pod{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 10}).
+		Complete(r)
+}
+
+type podTimingReconciler struct {
+	client                       client.Client
+	podEnterPhaseDurationSeconds *prometheus.GaugeVec
+	alreadyRecorded              sets.String
+	alreadyRecordedLock          sync.RWMutex
+}
+
+func (r *podTimingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	return ctrl.Result{}, r.reconcile(ctx, req)
+}
+
+func (r *podTimingReconciler) reconcile(ctx context.Context, req ctrl.Request) error {
+	if !strings.HasPrefix(req.Namespace, "e2e-clusters-") {
```

Comment:
TODO: label the namespaces we create, and set up the controller watch configuration to filter namespaces on label, then we shouldn't need this check